packages/app-builder-lib/src/targets/nsis/nsisUtil.ts

```diff
@@ -1,22 +1,41 @@
 import { Arch, log } from "builder-util"
 import { PackageFileInfo } from "builder-util-runtime"
-import { getBinFromUrl } from "../../binDownload"
+import { getBinFromUrl, getBinFromCustomLoc } from "../../binDownload"
 import { copyFile } from "builder-util/out/fs"
 import * as path from "path"
 import { getTemplatePath } from "../../util/pathManager"
 import { NsisTarget } from "./NsisTarget"
 import * as fs from "fs/promises"
 import * as zlib from "zlib"
+import { NsisOptions } from "./nsisOptions"
 
 export const nsisTemplatesDir = getTemplatePath("nsis")
 
+export const NsisTargetOptions = (() => {
+  let _resolve: (options: NsisOptions) => any
+  const promise = new Promise<NsisOptions>(resolve => (_resolve = resolve))
+  return {
+    then: (callback: (options: NsisOptions) => any): Promise<string> => promise.then(callback),
+    resolve: (options: NsisOptions): any => _resolve(options),
+  }
+})()
+
 export const NSIS_PATH = () => {
   const custom = process.env.ELECTRON_BUILDER_NSIS_DIR
   if (custom != null && custom.length > 0) {
     return Promise.resolve(custom.trim())
   }
-  // noinspection SpellCheckingInspection
-  return getBinFromUrl("nsis", "3.0.4.1", "VKMiizYdmNdJOWpRGz4trl4lD++BvYP2irAXpMilheUP0pc93iKlWAoP843Vlraj8YG19CVn0j+dCo/hURz9+Q==")
+  return NsisTargetOptions.then((options: NsisOptions) => {
+    if (options.customNsisBinary) {
+      const { checksum, url, version } = options.customNsisBinary
+      if (checksum && url) {
+        const binaryVersion = version || checksum.substr(0, 8)
+        return getBinFromCustomLoc("nsis", binaryVersion, url, checksum)
+      }
+    }
+    // noinspection SpellCheckingInspection
+    return getBinFromUrl("nsis", "3.0.4.2", "o+YZsXHp8LNihhuk7JsCDhdIgx0MKKK+1b3sGD+4zX5djZULe4/4QMcAsfQ+0r+a8FnwBt7BVBHkIkJHjKQ0sg==")
```

Comment:
@develar I tried to update to `3.0.5.0` that is uploaded in electron-builder-binaries, but that throws CI errors:
```
    Command line defined: "UNINSTALLER_OUT_FILE=C:\Users\RUNNER~1\AppData\Local\Temp\et-d8d6360ca6c99632f01b203b451e295c\t-oEMxqx\test-project-0\dist\__uninstaller-nsis-TestApp.exe"
    Processing config: C:\Users\runneradmin\AppData\Local\electron-builder\Cache\nsis\nsis-3.0.5.0\nsisconf.nsh
    Processing script file: "<stdin>" (ACP)

    Error output:
    Bad text encoding
    Error in macro MUI_LANGUAGEEX on macroline 13
    Error in macro MUI_LANGUAGE on macroline 4
    Error in macro addLangs on macroline 5
    Error in script "<stdin>" on line 112 -- aborting creation process
```

I couldn't easily identify why and `3.0.4.2` seems to have the log and max string length compiler flags enabled, so I'm just utilizing that here.

Reply:
@develar it happens in this code:
https://github.com/electron-userland/electron-builder-binaries/blob/ec807af771a341b93d122734a9a1d6d2ae9bf90a/nsis/Contrib/Modern%20UI%202/Localization.nsh#L29-L58

I think this is line 13? (Not sure how it's counted)
https://github.com/electron-userland/electron-builder-binaries/blob/ec807af771a341b93d122734a9a1d6d2ae9bf90a/nsis/Contrib/Modern%20UI%202/Localization.nsh#L50-L51

Alternatively, we could upgrade to makensis `3.0.6.1`, but I'm struggling to figure out how to compile it for both windows & mac on my hackintosh.